lib/shared/src/conversion.rs

```diff
@@ -0,0 +1,395 @@
+            "integer" | "int" => Ok(Conversion::Integer),
+            "float" => Ok(Conversion::Float),
+            "bool" | "boolean" => Ok(Conversion::Boolean),
+            "timestamp" => Ok(Conversion::Timestamp),
+            _ if s.starts_with("timestamp|") => {
+                let fmt = &s[10..];
+                // DateTime<Utc> can only convert timestamps without
+                // time zones, and DateTime<FixedOffset> can only
+                // convert with tone zones, so this has to distinguish
+                // between the two types of formats.
+                if format_has_zone(fmt) {
+                    Ok(Conversion::TimestampTZFmt(fmt.into()))
+                } else {
+                    Ok(Conversion::TimestampFmt(fmt.into()))
+                }
+            }
+            _ => Err(ConversionError::UnknownConversion { name: s.into() }),
+        }
+    }
+}
+
+/// Helper function to parse a conversion map and check against a list of names
+pub fn parse_check_conversion_map(
+    types: &HashMap<String, String>,
+    names: &[impl AsRef<str>],
+) -> Result<HashMap<String, Conversion>, ConversionError> {
+    // Check if any named type references a nonexistent field
+    let names = names.iter().map(|s| s.as_ref()).collect::<HashSet<_>>();
+    for name in types.keys() {
+        if !names.contains(name.as_str()) {
+            tracing::warn!(
+                message = "Field was specified in the types but is not a valid field name.",
+                field = &name[..]
+            );
+        }
+    }
+
+    parse_conversion_map(types)
+}
+
+/// Helper function to parse a mapping of conversion descriptions into actual Conversion values.
+pub fn parse_conversion_map(
+    types: &HashMap<String, String>,
+) -> Result<HashMap<String, Conversion>, ConversionError> {
+    types
+        .iter()
+        .map(|(field, typename)| {
+            typename
+                .parse::<Conversion>()
+                .map(|conv| (field.clone(), conv))
+        })
+        .collect()
+}
+
+#[derive(Debug, Eq, PartialEq, Snafu)]
+pub enum Error {
+    #[snafu(display("Invalid boolean value {:?}", s))]
+    BoolParseError { s: String },
+    #[snafu(display("Invalid integer {:?}: {}", s, source))]
+    IntParseError { s: String, source: ParseIntError },
+    #[snafu(display("Invalid floating point number {:?}: {}", s, source))]
+    FloatParseError { s: String, source: ParseFloatError },
+    #[snafu(display("Invalid timestamp {:?}: {}", s, source))]
+    TimestampParseError { s: String, source: ChronoParseError },
+    #[snafu(display("No matching timestamp format found for {:?}", s))]
+    AutoTimestampParseError { s: String },
+}
+
+impl Conversion {
+    /// Use this `Conversion` variant to turn the given `bytes` into a new `T`.
+    pub fn convert<T>(&self, bytes: Bytes) -> Result<T, Error>
+    where
+        T: From<Bytes> + From<i64> + From<f64> + From<bool> + From<DateTime<Utc>>,
+    {
```

Comment:
This method was updated to be generic over `T`, so that it no longer relies on `vector::Value`.